drivers/sdhc/sdhc_ambiq.c

```diff
@@ -0,0 +1,829 @@
+	}
+#endif /* CONFIG_PM_DEVICE_RUNTIME */
+
+	return ret;
+}
+
+/*
+ * Get Ambiq SDIO Host controller properties
+ */
+static int ambiq_sdio_get_host_props(const struct device *dev, struct sdhc_host_props *props)
+{
+	LOG_DBG("SDHC get host props");
+	memset(props, 0, sizeof(*props));
+	props->f_max = AMBIQ_SDIO_FREQ_MAX;
+	props->f_min = AMBIQ_SDIO_FREQ_MIN;
+	props->power_delay = 50; /* power_delay could be changed based on board power design */
+	props->host_caps.suspend_res_support = true;
+	props->host_caps.adma_2_support = true;
+	props->host_caps.sdio_async_interrupt_support = true;
+	props->host_caps.vol_180_support = true;
+	props->host_caps.bus_4_bit_support = true;
+	props->host_caps.bus_8_bit_support = true;
+	props->host_caps.high_spd_support = true;
+	props->host_caps.sdr50_support = true;
+	props->host_caps.sdr104_support = true;
+	props->host_caps.ddr50_support = true;
+	props->host_caps.hs200_support = true; /* Ambiq SDIO only supports 96MHz MAX frequency */
+	props->max_current_330 = 1020;
+	props->max_current_300 = 1020;
+	props->max_current_180 = 1020;
+	props->is_spi = false;
+	return 0;
+}
+
+/*
+ * Set I/O properties of SDIO host controller
+ */
+static int ambiq_sdio_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	struct ambiq_sdio_data *data = dev->data;
+	am_hal_host_bus_voltage_e eBusVoltage;
+	am_hal_host_bus_width_e eBusWidth;
+	am_hal_host_uhs_mode_e eUHSMode = AM_HAL_HOST_UHS_SDR50;
+	uint32_t ui32Status = 0;
+	int ret = 0;
+
+	LOG_DBG("(SDIO clock_freq=%d, bus_width=%d, timing_mode=%d, bus_mode=%d)", ios->clock,
+		ios->bus_width, ios->timing, ios->bus_mode);
+
+	if (ios->clock != 0 && (ios->clock <= AMBIQ_SDIO_FREQ_MAX) &&
+	    (ios->clock >= AMBIQ_SDIO_FREQ_MIN)) {
+		data->card.cfg.ui32Clock = ios->clock;
+	} else if (ios->clock != 0 && (ios->clock > AMBIQ_SDIO_FREQ_MAX) &&
+		   (ios->clock <= MMC_CLOCK_HS200)) {
+		/*Limit HS200 frequency to AMBIQ_SDIO_FREQ_MAX(96MHz) */
+		data->card.cfg.ui32Clock = AMBIQ_SDIO_FREQ_MAX;
+		eUHSMode = AM_HAL_HOST_UHS_SDR104;
+	} else if (ios->clock != 0) {
+		return -ENOTSUP;
+	}
+
+	if (ios->bus_mode != SDHC_BUSMODE_PUSHPULL) {
+		return -ENOTSUP;
+	}
+
+	switch (ios->bus_width) {
+	case SDHC_BUS_WIDTH1BIT:
+		eBusWidth = AM_HAL_HOST_BUS_WIDTH_1;
+		break;
+	case SDHC_BUS_WIDTH4BIT:
+		eBusWidth = AM_HAL_HOST_BUS_WIDTH_4;
+		break;
+	case SDHC_BUS_WIDTH8BIT:
+		eBusWidth = AM_HAL_HOST_BUS_WIDTH_8;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	switch (ios->signal_voltage) {
+	case SD_VOL_3_3_V:
+		eBusVoltage = AM_HAL_HOST_BUS_VOLTAGE_3_3;
+		break;
+	case SD_VOL_3_0_V:
+		eBusVoltage = AM_HAL_HOST_BUS_VOLTAGE_3_0;
+		break;
+	case SD_VOL_1_8_V:
+		eBusVoltage = AM_HAL_HOST_BUS_VOLTAGE_1_8;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	/*
+	 * Recommend to use 1.8V IO for SDIO Host controller.
+	 * External level shifter is required for voltage switch between 1.8V and 3V.
+	 */
+	if (ios->signal_voltage != SD_VOL_1_8_V) {
```

Comment:
Why is there a switch statement above if we are going to reject signal voltages besides 1.8V? Also, this won't work with the SD stack- cards start at 3.3V, and negotiate down to 1.8V